ocrd/validator/page_validator.py

```diff
@@ -0,0 +1,174 @@
+        Validates an OcrdPage for consistency
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        validator = PageValidator(page, **kwargs)
+        return validator.validate()
+
+
+    @staticmethod
+    def validate_ocrd_file(ocrd_file, **kwargs):
+        """
+        Validates an OcrdFile representing a PAGE doc for consistency
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        validator = PageValidator(from_file(ocrd_file), **kwargs)
+        return validator.validate()
+
+    def __init__(self, page, strictness='strict'):
+        """
+        Arguments:
+            strictness (string): 'strict', 'fix' or 'off'
+        """
+        self.report = ValidationReport()
+        self.page = page
+        self.strictness = strictness
+
+    def validate(self):
+        """
+        Do the validation / fixing.
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        if self.strictness == 'off':
+            return self.report
+        self.find_or_fix_inconsistencies(self.page, strictness=self.strictness)
+        return self.report
+
+    def concatenate_index_1_text(self, node, concatenate_with):
+        """
+        Concatenate children of node according to https://ocr-d.github.io/page#consistency-of-text-results-on-different-levels
+        """
+        _, _, getter, concatenate_with = [x for x in _HIERARCHY if isinstance(node, x[0])][0]
+        ret = ''
+        tokens = [self.get_index_1_text(x) for x in getattr(node, getter)()]
+        return concatenate_with.join(tokens).strip()
+
+    def get_index_1_text(self, node):
+        """
+        Get the most confident text results, either those with @index = 1 or the first text results or empty string.
+        """
+        textEquivs = node.get_TextEquiv()
+        if not textEquivs:
+            log.debug("No text results on %s %s", node, node.id)
+            return ''
+        else:
+            if len(textEquivs) > 1:
+                index1 = [x for x in textEquivs if x.index == 1]
+                if index1:
+                    return index1[0].get_Unicode().strip()
+            return textEquivs[0].get_Unicode().strip()
+
+    def set_index_1_text(self, node, text):
+        """
+        Set the most confident text results, either those with @index = 1, the first text results or add new one.
+        """
+        text = text.strip()
+        textEquivs = node.get_TextEquiv()
+        if not textEquivs:
+            node.add_TextEquiv(TextEquivType(Unicode=text))
+        else:
+            if len(textEquivs) > 1:
+                index1 = [x for x in textEquivs if x.index == 1]
+                if index1:
+                    index1[0].set_Unicode(text)
+                    return
+            textEquivs[0].set_Unicode(text)
+
+
+    def find_or_fix_inconsistencies(self, node, strictness='strict'):
```

Comment:
Split into two methods? `find_inconsistencies` returns a list of inconsistencies which can be fixed via `fix_inconsistencies`.

Reply:
It's a bit smelly that `report` is an instance property. Unfortunately that means that the whole instance has a state. Before implementing this in a PageValidator, I had the methods as functions in `ocrd_page`, where the behaviour you propose was easier.

I could add facades `find_inconsistencies` and `fix_inconsistencies` that only differ in what they accept as strictness?

Or just prepend an underscore and to make it clear that these methods are not part of the public API.

Reply:
The latter.

In understand that the distinction between `fix` and `find` depends on the `strictness` parameter, right? Maybe `handle_inconsistencies` could be an alternative?

Reply:
You mean `handle_inconsistencies` is clearer than `find_or_fix_inconsistencies`?

Reply:
called handle_inconsistencies now and again a function not a method (as are the former `*_index_1_*` methods).